presto-main/src/main/java/com/facebook/presto/server/protocol/QueryBlockingRateLimiter.java

```diff
@@ -0,0 +1,106 @@
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.ThreadPoolExecutor;
+import java.util.concurrent.TimeUnit;
+
+import static com.facebook.airlift.concurrent.Threads.daemonThreadsNamed;
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.util.concurrent.Futures.immediateFailedFuture;
+import static com.google.common.util.concurrent.Futures.immediateFuture;
+import static com.google.common.util.concurrent.MoreExecutors.listeningDecorator;
+import static java.util.Objects.requireNonNull;
+
+/*
+ * Rate Limiting per query with token bucket
+ * Rate = rateLimitBucketMaxSize/second
+ * When having sufficient tokens, Request will be responded immediately.
+ * When not having enough tokens available, it uses the delayed processing method.
+ */
+public class QueryBlockingRateLimiter
+{
+    private final LoadingCache<QueryId, RateLimiter> rateLimiterCache;
+    private final long rateLimitBucketMaxSize;
+    private final ListeningExecutorService listeningExecutorService;
+
+    private CounterStat rateLimiterTriggeredCounter = new CounterStat();
+
+    @Inject
+    public QueryBlockingRateLimiter(QueryManagerConfig queryManagerConfig)
+    {
+        requireNonNull(queryManagerConfig, "queryManagerConfig is null");
+        this.rateLimitBucketMaxSize = queryManagerConfig.getRateLimitBucketMaxSize();
+        checkArgument(rateLimitBucketMaxSize >= 1, "Rate limit token bucket max size should be at least 1");
+        // Using a custom thread pool with size 1-10 to reduce initial thread resources
+        ExecutorService executorService = new ThreadPoolExecutor(1, 10, 60, TimeUnit.SECONDS, new LinkedBlockingQueue<>(), daemonThreadsNamed("rate-limiter-listener"));
+        listeningExecutorService = listeningDecorator(executorService);
+        rateLimiterCache = CacheBuilder.newBuilder().maximumSize(10000).expireAfterAccess(60, TimeUnit.MINUTES).build(CacheLoader.from(key -> RateLimiter.create(rateLimitBucketMaxSize)));
+    }
+
+    /*
+     * For accidental bug-caused DoS, we will use delayed processing method to reduce the requests, even when user do not have back-off logic implemented
+     * Optimized to avoid blocking for normal usages with TryRequire first
+     * Fall back to delayed processing method to acquire a permit, in a separate thread pool
+     * @Return:time spent sleeping to enforce rate, in seconds; 0.0 if not rate-limited
+     */
+    public ListenableFuture<Double> acquire(QueryId queryId)
+    {
+        if (queryId == null) {
+            return immediateFailedFuture(new IllegalArgumentException("queryId should not be null"));
+        }
+        RateLimiter rateLimiter = rateLimiterCache.getUnchecked(queryId);
+        if (rateLimiter.tryAcquire()) {
+            return immediateFuture(0.0);
+        }
+        ListenableFuture<Double> asynTask = listeningExecutorService.submit(() -> rateLimiter.acquire());
```

Comment:
could this call wait for ever.?
Would it make sense to use tryAcquire with some sane timeout as a safeguard.?

Reply:
tryAcquire will return immediately, either true or false.

Reply:
There seems to be a flavor which does not immediately return. At least the timeout param [javadocs](https://guava.dev/releases/19.0/api/docs/com/google/common/util/concurrent/RateLimiter.html#tryAcquire(long,%20java.util.concurrent.TimeUnit)) seems to suggest that it will wait for the permit.
See below :

```
public boolean tryAcquire(long timeout, TimeUnit unit)
Acquires a permit from this RateLimiter if it can be obtained without exceeding the specified timeout, or returns false immediately (without waiting) if the permit would not have been granted before the timeout expired.
Parameters:
timeout - the maximum time to wait for the permit. Negative values are treated as zero.

```

Reply:
btw, I also tested the timeout logic; it does not really wait for that timeout, instead, it is using timeout to calculate whether to return true or false immediately.

Reply:
this approach sounds good then.